src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java

```diff
@@ -0,0 +1,390 @@
+                    }
+
+                    if(hasItem)
+                    {
+                        return true;
+                    }
+                }
+            }
+        }
+
+        return false;
+    }
+
+
+    /**
+     * Finds the first @see ItemStack the type of {@code is}.
+     * It will be taken from the chest and placed in the workers inventory.
+     * Make sure that the worker stands next the chest to not break immersion.
+     * Also make sure to have inventory space for the stack.
+     * @param entity the tileEntity chest or building.
+     * @param is the itemStack.
+     * @return true if found the stack.
+     */
+    public boolean isInTileEntity(TileEntityChest entity, ItemStack is)
+    {
+        return is != null
+                && InventoryFunctions
+                .matchFirstInInventoryWithInventory(
+                        entity,
+                        stack -> stack != null && is.isItemEqual(stack),
+                        InventoryFunctions::doNothing
+                );
+    }
+
+    @Override
+    public void readFromNBT(final NBTTagCompound compound)
+    {
+        super.readFromNBT(compound);
+    }
+
+    @NotNull
+    @Override
+    public NBTTagCompound writeToNBT(@NotNull final NBTTagCompound compound)
+    {
+        super.writeToNBT(compound);
+        return compound;
+    }
+
+    /**
+     * Dump the inventory of a citizen into the warehouse.
+     * Go through all items and search the right chest to dump it in.
+     * @param inventoryCitizen
+     */
+    public void dumpInventoryIntoWareHouse(@NotNull final InventoryCitizen inventoryCitizen)
+    {
+        for(int i = 0; i < inventoryCitizen.getSizeInventory(); i++)
+        {
+            ItemStack stack = inventoryCitizen.getStackInSlot(i);
+
+            if(stack == null || stack.getItem() == null || stack.stackSize == 0)
+            {
+                continue;
+            }
+            TileEntityChest chest = searchRightChestForStack(stack);
+            if(chest == null)
+            {
+                LanguageHandler.sendPlayersLocalizedMessage(getColony().getMessageEntityPlayers(), "com.minecolonies.coremod.wareHouse.full");
+                return;
+            }
+            InventoryUtils.addItemStackToInventory(chest, stack);
+            inventoryCitizen.removeStackFromSlot(i);
+        }
+
+    }
+
+    /**
+     * Search the right chest for an itemStack.
+     * @param stack the stack to dump.
+     * @return the tile entity of the chest
+     */
+    @Nullable
+    private TileEntityChest searchRightChestForStack(@NotNull final ItemStack stack)
+    {
+        if(InventoryUtils.findFirstSlotInInventoryWith(this, stack.getItem(), stack.getItemDamage()) != -1 && InventoryUtils.getOpenSlot(this) != -1)
+        {
+            return this;
+        }
+
+        for(BlockPos pos : getBuilding().getAdditionalCountainers())
+        {
+            TileEntity entity = worldObj.getTileEntity(pos);
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'entity' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)